### PR TITLE
Remove IteratorRange in favor of llvm::iterator_range

### PR DESCRIPTION
**Pull request by jrose-apple**
Now that `llvm::iterator_range` has `empty`, there's not enough reason to keep our own version of it in the Swift repo.

No functionality change.

### CODE REVIEW

**Comment by jrose-apple** [comment]
apple/swift-llvm#174
@swift-ci Please smoke test

File: include/swift/Basic/STLExtras.h [hunk 1]
```diff
@@ -251,37 +251,13 @@ inline Iterator prev_or_begin(Iterator it, Iterator begin) {
...
-template<typename Iterator>
-inline IteratorRange<Iterator> 
-makeIteratorRange(Iterator first, Iterator last) {
-  return IteratorRange<Iterator>(first, last);
-}
 
 /// An iterator that transforms the result of an underlying bidirectional
 /// iterator with a given operation.
 ///
+/// Slightly different semantics from llvm::map_iterator, but we should
```
Comment (gottesmm) [inline include/swift/Basic/STLExtras.h]: What are the different semantics here? Just seems out of place with the rest of the PR. Can you elaborate in the code?

File: include/swift/AST/DeclContext.h [hunk 2]
```diff
@@ -667,7 +667,7 @@ class DeclIterator {
 
 /// The range of declarations stored within an iterable declaration
 /// context.
-typedef IteratorRange<DeclIterator> DeclRange;
+typedef llvm::iterator_range<DeclIterator> DeclRange;
```
Comment (gottesmm) [inline include/swift/AST/DeclContext.h]: `using`?

**Review by gottesmm — APPROVED** [review]
Overall looks good to me. Two small nits and a request. We use this a lot. Why not inject this into LLVM.h and strip the llvm::?

File: include/swift/AST/DeclContext.h (hunk 2 above)
Comment (jrose-apple) [reply include/swift/AST/DeclContext.h]: I mostly wasn't updating existing code, but sure.

File: include/swift/Basic/STLExtras.h (hunk 1 above)
Comment (jrose-apple) [reply include/swift/Basic/STLExtras.h]: It's something about whether the underlying iterator returns a reference or not. I don't actually remember the details, but I tried to remove it and failed and I want to leave a warning for others to come.

**Comment by jrose-apple** [comment]
> Why not inject this into LLVM.h and strip the llvm::?

Sure, will do.

File: include/swift/Basic/STLExtras.h (hunk 1 above)
Comment (jrose-apple) [reply include/swift/Basic/STLExtras.h]: It is *not* the purpose of the PR, as you note, so if you object to it going in in this form I'll just drop it.

**Comment by jrose-apple** [comment]
apple/swift-llvm#174
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
Hm, guess I have to wait until the LLVM change propagates into `stable`.

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test macOS

**Pull request merged**